### PR TITLE
Hybrid Model: Zone Infiltration and People Count

### LINKED ISSUE #7026: Documentation Issues - 2018b
Issue overview
--------------
A collection of small documentation issues. See checkboxed comments below.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by xuanluo113**
Pull request overview
---------------------

### 1. Fix #7026 (original pull request #7130) issues in HybridModel:Zone, including:
- [x] For example file HybridZoneModel.idf with the request for Output:VariableDictionary, "Zone Infiltration Hybrid Model Air Change Rate" is not in the RDD, while the Flag should be initialized here since there are objects that have Calculate Zone Internal Thermal Mass or Calculate Zone Air Infiltration Rate enabled.
- [x] This object could read begin/end day and month from the RunPeriod.

**Input-output-reference:**
- [x] Section heading for the object `HybridModel:Zone`.
- [x] The overview description of the feature is not clear, it could use some additional text. e.g. "for existing buildings _by supplying measured temperature data_."
- [x] Sample idf object.
- [x] Outputs section.
- [x] Most of the field descriptions need work. The Yes/No fields don't mention "No" as an option, and they don't indicate what the default is.
- [ ] The Year/Month field descriptions are all wrong. "... for the annual run period desired." should be "... for the period of the measured data during which the hybrid model will be active." (or something similar)
- [x] The description of the temperature schedule says to reference a Schedule:File object, but does the code restrict this to Schedule:File or can it be any schedule object? The description here needs to make that clear. And it also needs to mention that all schedules must include data for a full year, so especially when using schedule:file, the dates outside the range of the actual measured data can be zeros, but all of the rows must be there or the data will be read for the wrong dates. Unless the hybrid code is treating this schedule in a special way.

### 2. Add zone infiltration and people count to hybrid modeling

The proposed new feature improves the current HybridModel modeling capability (Lee & Hong, 2017) by adding inverse modeling algorithms based on the zone moisture and contaminant balance equations. The current HybridModel feature uses measured zone air temperature to inversely solve the zone internal thermal mass or air infiltration rate assuming the HVAC system is off and zone is in free-floating mode. The new HybridModel feature allows additional parameters such as the measured zone air humidity ratio and CO<sub>2</sub> concentration as the input to the inverse modeling algorithms to solve the highly uncertain parameters such as air infiltration and people count. Depending on the
availability of the measured zone supply air parameters (i.e., supply air flow rate, supply air temperature, supply air humidity ratio, and supply air CO<sub>2</sub> concentration), the new HybridModel feature can solve air infiltration and people count for both HVAC-off mode and HVAC-on
mode. The hybrid approach keeps the virtue of the physics-based model and takes advantage of more measured buildings data which become available nowadays due to the wide adoption of low-cost sensors and the needs of better controls in existing buildings. This new feature proposal provides technical details of the enhanced HybridModel object and its implementation in EnergyPlus. The proposed feature will improve simulation usability and accuracy for existing buildings, which supports
more accurate analysis of energy retrofits.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Functional code review (it has to work!)
 - n/a If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [x] Performance: CI Linux results include performance check
 - [x] Unit Test(s)
 - C++ checks:
   - [x] Argument types
   - [x] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [x] Verify naming conventions and styles, memos and notes and defaults
   - [x] Open windows IDF Editor with modified IDD to check for errors
   - [x] If transition, add to input rules file for interfaces
   - [x] If transition, add transition source
   - [x] If transition, update idfs
 - [x] If new idf included, locally check the err file and other outputs
 - [x] Required documentation updates?
 n/a ExpandObjects changes?
 n/a If output changes, including tabular output structure, add to output rules file for interfaces

Remaining Items to Resolve
------------------------------
### Must-Haves Before I/O Freeze
- [x] Transition rules
- [x] Transition code
- [x] New example files version should be 9.1
- [x] Unit test segfaults
- [x] Unit test failures
- [x] Unused variable build warnings
- [x] Incorrect formatting [build warnings](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-1a5b60a0bfce0307b6f5c4cd63769fa90d6bab99-x86_64-Linux-Ubuntu-18.04-custom_check.html)
### Others
- [x] Keep 5 new example files or consolidate into fewer or move some to _filename for internal testing only?
- [x] Minor doc edits
- [x] New example files - filename comments at top need to be fixed
- [x] Clean up warnings in example files

### CODE REVIEW

**Comment by mjwitte** [comment]
@XuanLuoLBL I've reviewed the NFP.
1. The current IDD says that "Zone Air Infiltration Rate and Zone Internal Thermal Mass calculations cannot be performed simultaneously". Will that still be the case here, that only one of Thermal Mass, Infiltration, or People can be actively calculated? Do you anticipate being able to do more than one simultaneously in the future?
2. I cannot find HybridModel:Zone in the current I/O Ref (?!).
3. Remind me why this object has date range fields? Should these be expanded to include Begin Year and End Year, similar to the updated RunPeriod object?

**Comment by tsbyq** [comment]
@mjwitte Thanks Michael for the comments, here are the replies:

1. The current version only calculates one of the three unknown parameters: zone air infiltration rate, people count, and thermal mass at a time. Yes, solving two or more unknown parameters with more measured parameters at the same time will be future considerations.

2. We are working on the IO reference.

3. It is inherited from the original HybridModel:Zone object. But we'll check that.

**Comment by jasondegraw** [comment]
@XuanLuoLBL @tsbyq A question in addition to @mjwitte's: How is this object (and HybridModel:Zone) expected to behave with the begin/end dates specified in a multiyear simulation?

**Comment by nrel-bot-3** [comment]
@XuanLuoLBL @lgentile it has been 28 days since this pull request was last updated.

File: src/EnergyPlus/InternalHeatGains.cc
Comment (mjwitte) [inline src/EnergyPlus/InternalHeatGains.cc:3779]: Why are these changes in here?

File: src/EnergyPlus/InternalHeatGains.cc
Comment (mjwitte) [inline src/EnergyPlus/InternalHeatGains.cc]: Use `ZoneIntGain(ZoneNum).Device(IntGainsNum).CompTypeOfNum != IntGainTypeOf_People` integer, instead of (slow) string compare. Here and in `SumAllInternalCO2GainsExceptPeople`.

File: src/EnergyPlus/InternalHeatGains.hh
Comment (mjwitte) [inline src/EnergyPlus/InternalHeatGains.hh:100]: End comment does not apply here - this function remains unchanged and isn't only for hybridmodel.

File: idd/Energy+.idd.in
Comment (mjwitte) [inline idd/Energy+.idd.in]: In all of the notes, delete the A6 etc. references. Just the fields names are enough. I trust that all of these rules are trapped in the input processing?

File: src/EnergyPlus/ZoneContaminantPredictorCorrector.cc
Comment (mjwitte) [inline src/EnergyPlus/ZoneContaminantPredictorCorrector.cc]: Why add this here? `GetHybridModelZone()` is already called in `ZoneTempPredictorCorrector::GetZoneAirSetPoints`  and there is no logic in `GetHybridModelZone()`  to prevent it from processing the input twice.  Seems like the call from `ZoneTempPredictorCorrector::GetZoneAirSetPoints` should be adequate.

File: src/EnergyPlus/ZoneTempPredictorCorrector.cc
Comment (mjwitte) [inline src/EnergyPlus/ZoneTempPredictorCorrector.cc:5136]: It would be cleaner to move the entire hybrid model section to a separate function. Is it using very many local variables from `CorrectZoneAirTemp`? Instead of this complex if, this entire block (or the new call to a separate function) should be wrapped with `if HybridModel::FlagHybridModel`. No need to evaluate the rest if this is false. Similar comments for predictor, and for contaminant predictor and corrector.

File: src/EnergyPlus/HybridModel.hh
Comment (mjwitte) [inline src/EnergyPlus/HybridModel.hh]: Nitpicking - these three variables should be `People...`.

File: src/EnergyPlus/InternalHeatGains.cc
Comment (mjwitte) [inline src/EnergyPlus/InternalHeatGains.cc]: Delete? Appears to be unused.

File: testfiles/CMakeLists.txt
Comment (mjwitte) [inline testfiles/CMakeLists.txt]: 5 new example files - that's a lot. @Myoldmopar do we want to move any of these to start with underscore (not included in distribution)? Or perhaps combine these such that each zone illustrates a different method?

File: tst/EnergyPlus/unit/ZoneTempPredictorCorrector.unit.cc
Comment (mjwitte) [inline tst/EnergyPlus/unit/ZoneTempPredictorCorrector.unit.cc:299]: Not necessary here. `HybridModel::clear_state()` will do this. In fact, most (all?) of these deallocates here aren't necessary.

**Review by mjwitte — CHANGES_REQUESTED** [review]
1. There are build warnings.
2. integration.HybridZoneModel fails with fatal errors.
3. This requires transition rules in src/Transition/InputRulesFiles/Rules9-0-0-to-9-1-0.md (see earlier ones for examples).
4. Will you add transition source code or do you need me or @Myoldmopar to do that?
5. Needs units tests for the new features.

I have not reviewed docs yet - probably not until tomorrow.

File: src/EnergyPlus/ZoneContaminantPredictorCorrector.cc
Comment (tsbyq) [reply src/EnergyPlus/ZoneContaminantPredictorCorrector.cc]: The GetHybridModelZone() was moved to ZoneTempPredictorCorrector::GetZoneAirSetPoints in the latest commit.

File: doc/input-output-reference/src/overview/group-hybrid-model.tex
Comment (mjwitte) [inline doc/input-output-reference/src/overview/group-hybrid-model.tex]: \sub~sub~section
And this heading should be before the above paragraph, right after the Group heading.

File: doc/input-output-reference/src/overview/group-hybrid-model.tex
Comment (mjwitte) [inline doc/input-output-reference/src/overview/group-hybrid-model.tex:9]: All occurrences of \paragraph should be followed by a blank line so the text below starts a new paragraph.

File: doc/input-output-reference/src/overview/group-hybrid-model.tex
Comment (mjwitte) [inline doc/input-output-reference/src/overview/group-hybrid-model.tex]: When referring to other field names, omit "Field:" it just clutters things up.

**Review by mjwitte — CHANGES_REQUESTED** [review]
Skimmed through the Engineering Ref section looking for obvious formatting problems. There are some *impressive* equations in this section!

A few comments on the I/O Ref are noted.

There are still build warnings, seg faults, and unit test failures. These need to be addressed.

**Comment by mjwitte** [comment]
@XuanLuoLBL Merged in develop, fixed formatting problems (tabs!), and added transition rules. Working on transition source code now.

**Comment by mjwitte** [comment]
Added transition code and (hopefully) fixed test failures.

**Comment by mjwitte** [comment]
@XuanLuoLBL Running unit tests locally here, `EnergyPlusFixture.ZoneTempPredictorCorrector_CorrectZoneAirTempTest` fails. Others may fail, too, so watch here for CI report.

**Comment by xuanluo113** [comment]
@mjwitte Working on this, along with warnings, reduce the number of testfiles, add more unit tests and address other comments. Will pull and test and commit soon.

**Comment by mjwitte** [comment]
@XuanLuoLBL Good. Please see checklist at the top of this PR.

**Comment by jasondegraw** [comment]
@Myoldmopar @mjwitte @XuanLuoLBL I'm going to defer to others on this PR, but the continued presence of the start/end dates is a higher level problem that is going to need to be cleared up at some point. At the very least, the role of years needs to be clarified so that it fits in with the rest of the code. I don't want to hold up this PR for that though, so could we get an issue submitted to clear all this up?

**Comment by xuanluo113** [comment]
@jasondegraw @tsbyq We have modified the I/O ref and claimed that the Start and End Day and Month are only "for the period of the measured data during which the hybrid model will be active." And the feature is not yet able to handle importing mulpitle years of measured data. Shall we change the IDD naming to Measured Data Start Day, Month, etc.?

> @Myoldmopar @mjwitte @XuanLuoLBL I'm going to defer to others on this PR, but the continued presence of the start/end dates is a higher level problem that is going to need to be cleared up at some point. At the very least, the role of years needs to be clarified so that it fits in with the rest of the code. I don't want to hold up this PR for that though, so could we get an issue submitted to clear all this up?

**Comment by jasondegraw** [comment]
@XuanLuoLBL  It's a start, but what happens when a multiyear run is attempted (by which I mean there's a RunPeriod with multiple years in it, no changes to the hybrid model inputs)? What's supposed to happen? Are leap years handled correctly? I could go on with the questions, that's why I'm suggesting a new issue to clean this up later. In any case, I'll look back at the IORef in case I missed something but the code doesn't look like it is very robust.

**Comment by mjwitte** [comment]
@XuanLuoLBL Cleaned up (almost all) remaining testfile warnings and fixed the I/O ref headers (one more time). Will merge after CI completes one more round.

**Comment by mjwitte** [comment]
Unexpected eio/eso/shd diffs are unrelated to this PR. Noted in https://github.com/NREL/EnergyPlus/issues/7176.

**Comment by mjwitte** [comment]
Actually, the test files for this *really are* in relative coordinates, reverted the zone origin changes and correct the geometry rules. Hopefully this will be the last CI round.

**Review by mjwitte — APPROVED** [review]

**Comment by mjwitte** [comment]
@XuanLuoLBL Now that this is in `develop`, diffs are failing on the two new files, because there are duplicate column headers in the eso/csv output. This indicates that `SetUpOutputVariable` is executing twice for these variables .
image

**Pull request merged**
